### PR TITLE
[clang][Sema] Don't emit 'declared here' note for builtin functions with no decl in source

### LINKED ISSUE #93369: "did you mean [builtin function]?" diagnostics print a useless declaration site
Consider:
```c++
constexpr float A = __builtin_isinfinity();
```

The output is:
```console

array.cpp:1570:21: error: use of undeclared identifier '__builtin_isinfinity'; did you mean '__builtin_isfinite'?
 1570 | constexpr float A = __builtin_isinfinity();
      |                     ^~~~~~~~~~~~~~~~~~~~
      |                     __builtin_isfinite
array.cpp:1570:21: note: '__builtin_isfinite' declared here
array.cpp:1570:42: error: too few arguments to function call, expected 1, have 0
 1570 | constexpr float A = __builtin_isinfinity();
      |                     ~~~~~~~~~~~~~~~~~~~~~^
2 errors generated.
```

The "note: '__builtin_isfinite' declared here" line is useless and confusing - It's not declared there at all and in fact is declared nowhere since it's a builtin function.

We should simply not be adding this note for builtin functions.

### PR DESCRIPTION
**Pull request by JOE1994**
Fixes #93369

### CODE REVIEW

File: clang/lib/Sema/SemaLookup.cpp [hunk 1]
```diff
@@ -5897,6 +5897,16 @@ void Sema::diagnoseTypo(const TypoCorrection &Correction,
 
   NamedDecl *ChosenDecl =
       Correction.isKeyword() ? nullptr : Correction.getFoundDecl();
+
+  // For builtin functions which aren't declared anywhere in source,
+  // don't emit the "declared here" note.
+  if (auto *FD = dyn_cast_or_null<FunctionDecl>(ChosenDecl);
```
Comment (tbaederr) [inline clang/lib/Sema/SemaLookup.cpp:5903]:
```suggestion
  if (const auto *FD = dyn_cast_if_present<FunctionDecl>(ChosenDecl);
```

File: clang/test/SemaCXX/invalid-if-constexpr.cpp [hunk 2]
```diff
@@ -5,7 +5,7 @@ void similar() { // expected-note {{'similar' declared here}}
   if constexpr (similer<>) {} // expected-error {{use of undeclared identifier 'similer'; did you mean 'similar'?}}
 }
 void a() { if constexpr (__adl_swap<>) {}} // expected-error{{use of undeclared identifier '__adl_swap'; did you mean '__sync_swap'?}} \
-                                           // expected-note {{'__sync_swap' declared here}}
+                                           // not-expected-note {{'__sync_swap' declared here}}
```
Comment (tbaederr) [inline clang/test/SemaCXX/invalid-if-constexpr.cpp:8]: I don't think the `not-` prefix actually does anything with `-verify`, does it?

**Comment by tbaederr** [comment]
Would be nice if you could add new explicit tests for this.

File: clang/test/SemaCXX/invalid-if-constexpr.cpp (hunk 2 above)
Comment (JOE1994) [reply clang/test/SemaCXX/invalid-if-constexpr.cpp]: You are right!

File: clang/lib/Sema/SemaLookup.cpp [hunk 3]
```diff
@@ -5897,6 +5897,16 @@ void Sema::diagnoseTypo(const TypoCorrection &Correction,
 
   NamedDecl *ChosenDecl =
       Correction.isKeyword() ? nullptr : Correction.getFoundDecl();
+
+  // For builtin functions which aren't declared anywhere in source,
+  // don't emit the "declared here" note.
+  if (const auto *FD = dyn_cast_if_present<FunctionDecl>(ChosenDecl);
+      FD && FD->getBuiltinID() &&
+      PrevNote.getDiagID() == diag::note_previous_decl &&
+      Correction.getCorrectionRange().getBegin() == FD->getBeginLoc()) {
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaLookup.cpp:5906]: I wonder if checking location is necessary at all?

**Review by Fznamznon — COMMENTED** [review]
Could you please add a release note?

File: clang/lib/Sema/SemaLookup.cpp (hunk 3 above)
Comment (JOE1994) [reply clang/lib/Sema/SemaLookup.cpp:5906]: This location checking was motivated by test failure of `Clang :: Sema/implicit-decl.c`.

https://github.com/llvm/llvm-project/blob/8760d4ba4cb12d5cac2469f26cd09a2b3acd3c50/clang/test/Sema/implicit-decl.c#L13

I added this check to let Clang emit a **"declared here"** note for a "builtin" function (one with non-zero BuiltinID) that has an explicit declaration.

**Pull request merged**